### ISSUE TITLE
Real plugin architecture & clean installation flow

### ISSUE THREAD

**Issue by AntoineThebaud** [perses/perses#1543]
### Proposal

Our current plugin architecture technically works, still it's tightly coupled with Prometheus datasource and installing a new plugin is not possible as is (it's possible to hot-load the backend part of it, but for the frontend part doing 99% of the job it actually has to be part packaged with Perses).

The idea would be to really have externalized plugins & provide a single entry point to install them, that would take as input a package containing both the backend & frontend parts of the plugin, be able to "install" it & have Perses updated live accordingly, with as less amount of actions as possible:
- using the UI, ideally drag & drop + submit, that's it?
- using the CLI, ideally single command?
- provisioned set of plugins when starting Perses that would work using the same package format?

**Comment by AntoineThebaud** [comment]
FYI starting a POC around this topic here https://github.com/perses/poc-plugin-v2

**Comment by jgbernalp** [comment]
In addition to this:

What pain points did we have?
- Too many coupled providers make the plugin usage cumbersome when using them in embedding environments and also in Perses UI, maybe creating more reusable components and let the consumer, including Perses UI, connect them with providers can be more flexible.
- Although There is inherent coupling between queries, datasources and variables. In the UI the panel plugins and the queries are also coupled, this creates the challenge to adjust every panel to a specific query/datasource. Maybe query or datasource plugins should generate a common format (DataFrame) so they can be connected with several panels without a re-implementation
- (same as point 1) The plugin needs to be in the core source code, so it's not really pluggable at runtime.

What would an easy plugin usage look like?
- The plugin(s) should be self contained as a single folder / file that can be installed on runtime or pre-packaged in Perses
- Panel plugin components could be dependent only on ECharts easing the embedded usage, reducing the footprint the reusable components should have.
- Have a Perses plugin SDK (point 5) so boilerplates can be generated and packaged for distribution.
- If plugins can be added at runtime we need to make sure we isolate their execution (sandboxing) so an error wont crash the UI / Backend, in the frontend this is already considered using error boundaries.
- We can define a PluginManager that takes care of the discovery, installation and plugin lifecycle: installation, activation, deactivation, and cleanup.
- It would be great to have an official plugin marketplace/catalog where users can discover and install plugins easily

**Comment by Nexucis** [comment]
on top of what have been said, I think it will be cool if we are able to include also this issue https://github.com/perses/perses/issues/1546.

Would be cool to be able to provide an icon in the plugin package !

**Comment by jgbernalp** [comment]
Extending the work that @AntoineThebaud made for webpack
- https://github.com/perses/poc-plugin-v2/pull/1

 I included:

- An rspack variant for dynamic federated modules: https://github.com/perses/poc-plugin-v2/pull/2
- A Perses plugin npm create command: https://github.com/perses/poc-plugin-v2/pull/3

**Comment by AntoineThebaud** [comment]
Discussion about the plugin architecture is ongoing here https://github.com/perses/perses/discussions/2085

**Comment by AntoineThebaud** [comment]
Just to link this issue to the work done since:
- the code of the "official" plugins in the new architecture is hosted in a separated repo, [perses/plugins](https://github.com/perses/plugins).
- the adaptation of Perses' code to the new architecture is currently done on the `plugin-runtime` feature branch. See full list of PRs with this search https://github.com/perses/perses/pulls?q=is%3Apr+base%3Afeature%2Fplugin-runtime

**Comment by Nexucis** [comment]
Reviving a bit this issue to give a state on where we are today.

We are really close to be able to merge the feature branch `feature/plugin-runtime` to the `main` branch.
Before doing that we need to merge : 
- https://github.com/perses/perses/pull/2557. This PR is loading the new form from the plugins and is overriding the old plugin system.
- https://github.com/perses/plugins/pull/25. This PR is migrating all existing plugin outside of `perses/perses`
- https://github.com/perses/perses/pull/2567. This one is loading the `migrate.cue` files from the external plugins.

Once everything is merged, it will also mean that all plugins have been successfully migrated to `perses/plugins` and we will be able to cleanup the repo `perses/perses`.

Of course the work doesn't stop there. Here a list of the work that remains to do after the new plugin system is merged into the `main` branch (I am certainly missing other points. @jgbernalp @AntoineThebaud  feel free to complete this list):

- [ ] Remove all schema and migrate.cue files
- [ ] Migrate the go files representing the plugin to `perses/plugin`
- [ ] Remove package `@perses-dev/prometheus`
- [ ] Remove package `@perses-dev/tempo`
- [ ] Probably cleanup the package `@perses-dev/components` and `@perses-dev/core`
- [ ] Rework CLI command `plugin lint` and `plugin build` based on what have been done in the Perses backend in term of verification. We are checking more files, more structs that currently done in `perses/plugins`. Basically `plugin lint` should verify that all files required for the plugins are there with the required info. It should also check in `package.json` that the struct describing the plugin is present.
- [ ] When developing a plugin, it will be great that Perses is able to load it. So you can see during the development of your plugin the result directly in the UI.
- [ ] We will certainly have to put back in place a plugin watcher to reload the plugin list at runtime. It will be certainly required to make the plugin development environment working

**Comment by Nexucis** [comment]
I tried to remove the package `@perses-dev/prometheus`, but it's actually directly used in the explorer. So I guess we will have to find a way to load plugin in the explorer and stop doing specific things.

/cc @AntoineThebaud @Gladorme @jgbernalp  @celian-garcia

**Comment by celian-garcia** [comment]
> I tried to remove the package `@perses-dev/prometheus`, but it's actually directly used in the explorer. So I guess we will have to find a way to load plugin in the explorer and stop doing specific things.
> 
> /cc @AntoineThebaud @Gladorme @jgbernalp  @celian-garcia 

Indeed, I remember that we talked about it without entering into details.
The Prometheus explorer is something deeply linked with Prometheus. This is in my opinion not something that you can generalize for all time series databases. 
To me you need to keep some Prometheus stuff in Perses codebase. Everything that is not about ts plugin, and call it if the Prometheus plugin is enabled.

**Comment by Gladorme** [comment]
Yep, explorer has been partially hardcoded for metric / trace. I think plugin should be able to add new tabs in the explorer and plugin is owner of the content 😄

**Comment by jgbernalp** [comment]
Yes this probably can be its own plugin type `Explorer`. I'll take a look to use a plugin here.

**Comment by Nexucis** [comment]
Yeah I guess this is the way. That would also introduces the first plugin with no schema. 

A good thing to be able to transfer / support what @celian-garcia did with the snow plugin

**Comment by Nexucis** [comment]
One thought about the `Explorer` plugin, I am wondering how much it can be shared. For example, once you have a ClickHouse datasource, you would like to use it to explore your traces or metrics.
And it will be super similar than the Tempo explorer we have. Just the query will change.

Or we said we don't care because each explorer bringing by each datasources are super specific like the one for Prometheus with PromLens. So trying to share an explorer does not make sense or it's too much work for a not obvious result

**Comment by celian-garcia** [comment]
> Yeah I guess this is the way. That would also introduces the first plugin with no schema. 
> 
> A good thing to be able to transfer / support what @celian-garcia did with the snow plugin 

Well, mine is just to add a button in the menu where the explorer is a new tab in the explorer but I guess there will be similitudes indeed. I will revive my PR before Christmas I promise :D

**Comment by jgbernalp** [comment]
I created 2 WIP PRs to include the new "Explore" plugin: 

- https://github.com/perses/perses/pull/2598

- https://github.com/perses/plugins/pull/37

I think explorer plugins can bring whatever they need to explore a specific datasource or add support for several query types. Nothing stop us form creating a Explore plugin for "TimeSeriesQuery" and "TraceQuery" that will list the available datasources that provide time series. For now I moved the prometheus and tempo implementations to their respective plugins.

**Comment by AntoineThebaud** [comment]
FYI the last part of the backend adaptation to the v2 of the plugin architecture concerns Cuelang and is tracked with a dedicated issue: https://github.com/perses/perses/issues/1984

**Comment by Nexucis** [comment]
> We will certainly have to put back in place a plugin watcher to reload the plugin list at runtime. It will be certainly required to make the plugin development environment working

The more I am looking at this line, the more I am really wondering it it will be necessary. Loading at runtime the plugin doesn't seem realistic in a production environment as you will likely won't be able to download plugin or to add them on the flow.

**Comment by Nexucis** [comment]
This issue can be closed as everything required for the new plugin system has been done and we are reaching the final and stable release v0.51
